jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java

```diff
@@ -0,0 +1,489 @@
+                this.family = family;
+                this.srcIP = srcIP;
+                this.srcPort = srcPort;
+                this.dstIP = dstIP;
+                this.dstPort = dstPort;
+            }
+
+            public String getFamily()
+            {
+                return family;
+            }
+
+            public String getSourceAddress()
+            {
+                return srcIP;
+            }
+
+            public int getSourcePort()
+            {
+                return srcPort;
+            }
+
+            public String getDestinationAddress()
+            {
+                return dstIP;
+            }
+
+            public int getDestinationPort()
+            {
+                return dstPort;
+            }
+
+            @Override
+            public ClientConnectionFactory apply(ClientConnectionFactory factory)
+            {
+                return new V1(factory);
+            }
+
+            @Override
+            public boolean equals(Object obj)
+            {
+                if (this == obj)
+                    return true;
+                if (obj == null || getClass() != obj.getClass())
+                    return false;
+                Info that = (Info)obj;
+                return family.equals(that.family) &&
+                    Objects.equals(srcIP, that.srcIP) &&
+                    srcPort == that.srcPort &&
+                    Objects.equals(dstIP, that.dstIP) &&
+                    dstPort == that.dstPort;
+            }
+
+            @Override
+            public int hashCode()
+            {
+                return Objects.hash(family, srcIP, srcPort, dstIP, dstPort);
+            }
+        }
+    }
+
+    public static class V2 extends ProxyProtocolClientConnectionFactory
+    {
+        public V2(ClientConnectionFactory factory)
+        {
+            super(factory);
+        }
+
+        @Override
+        protected ProxyProtocolConnection newProxyProtocolConnection(EndPoint endPoint, Map<String, Object> context)
+        {
+            HttpDestination destination = (HttpDestination)context.get(HttpClientTransport.HTTP_DESTINATION_CONTEXT_KEY);
+            Executor executor = destination.getHttpClient().getExecutor();
+            V2.Info info = (V2.Info)destination.getOrigin().getTag();
+            if (info == null)
+            {
+                InetSocketAddress local = endPoint.getLocalAddress();
+                InetSocketAddress remote = endPoint.getRemoteAddress();
+                boolean ipv6 = remote.getAddress() instanceof Inet6Address;
+                info = new V2.Info(Info.Command.PROXY, ipv6 ? Info.Family.INET6 : Info.Family.INET4, Info.Protocol.STREAM, local.getHostString(), local.getPort(), remote.getHostString(), remote.getPort());
+            }
+            return new ProxyProtocolConnectionV2(endPoint, executor, getClientConnectionFactory(), context, info);
+        }
+
+        public static class Info implements ClientConnectionFactory.Decorator
```

Comment:
Shouldn't the V2 Info class extend the V1 Info class?  Or better yet just have a single Info class that can be used for both and that has a better name than `Info`.